File: Frontend/src/utils/validaciones.js
/**
 * Utilidades de validación compartidas para todos los módulos.
 * Reglas enfocadas en el contexto colombiano.
 */

// ─────────────────────────────────────────────────────────────────────────────
// TRANSPORTADORAS COLOMBIANAS RECONOCIDAS
// ─────────────────────────────────────────────────────────────────────────────
export const TRANSPORTADORAS_CO = [
  "Servientrega",
  "Interrapidísimo",
  "Inter Rapidísimo",
  "472",
  "DHL",
  "FedEx",
  "Coordinadora",
  "TCC Expreso",
  "TCC",
  "Envía",
  "Deprisa",
  "La Lleva",
  "Mensajería Urbana",
  "Lógistica",
  "Domicilios Nacional",
  "Speed Courier",
  "Adpostal",
];

// ─────────────────────────────────────────────────────────────────────────────
// HELPERS INTERNOS
// ─────────────────────────────────────────────────────────────────────────────

/** Solo letras (incl. tildes y ñ), espacios y guiones */
const REGEX_SOLO_LETRAS = /^[a-zA-ZáéíóúüñÁÉÍÓÚÜÑ\s'-]+$/;

/**
 * Comprueba que el texto tenga "sentido" en español:
 * - Al menos el 15 % de los caracteres alfabéticos son vocales.
 * - Bloquea "qwerty", "asdfgh", secuencias aleatorias.
 */
const tieneSentido = (str) => {
  const letras = str.replaceAll(/[^a-zA-ZáéíóúüñÁÉÍÓÚÜÑ]/g, "");
  if (letras.length < 2) return false;
  const vocales = (letras.match(/[aeiouáéíóúü]/gi) || []).length;
  return vocales / letras.length >= 0.15;
};

// ─────────────────────────────────────────────────────────────────────────────
// FILTROS DE ENTRADA EN TIEMPO REAL (para usar en onChange / onInput)
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Filtra caracteres no permitidos en un campo de número de documento.
 *  - esPasaporte = true  → permite dígitos + máximo 2 letras (mayúsculas).
 *  - esPasaporte = false → solo dígitos.
 * Devuelve el valor filtrado para asignar al estado del input.
 */
export const filtrarInputDocumento = (valor, esPasaporte = false) => {
  if (esPasaporte) {
    // Solo alfanuméricos
    const limpio = valor.replace(/[^a-zA-Z0-9]/g, "");
    let letrasCount = 0;
    let resultado = "";
    for (const ch of limpio) {
      if (/[a-zA-Z]/.test(ch)) {
        letrasCount++;
        if (letrasCount <= 2) resultado += ch.toUpperCase();
      } else {
        resultado += ch;
      }
    }
    return resultado;
  }
  // Solo dígitos
  return valor.replace(/\D/g, "");
};

/**
 * Filtra caracteres no permitidos en un campo de nombre / apellido.
 * Solo permite letras (incl. tildes y ñ), espacios, apóstrofos y guiones.
 */
export const filtrarInputNombre = (valor) =>
  valor.replace(/[^a-zA-ZáéíóúüñÁÉÍÓÚÜÑ\s'-]/g, "");

// ─────────────────────────────────────────────────────────────────────────────
// NOMBRES / APELLIDOS
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Valida un nombre o apellido.
 * @returns {string|null} Mensaje de error, o null si es válido.
 */
export const validarNombre = (str) => {
  const s = (str || "").trim();
  if (!s || s.length < 2) return "El nombre debe tener al menos 2 caracteres.";
  if (!REGEX_SOLO_LETRAS.test(s))
    return "El nombre solo puede contener letras, espacios y guiones. No se permiten números ni caracteres especiales.";
  if (!tieneSentido(s))
    return "El nombre no parece válido. Evite escribir letras al azar o sin sentido.";
  return null;
};

/**
 * Valida un nombre completo (puede contener 2 tokens: nombre + apellido).
 */
export const validarNombreCompleto = (str) => {
  const s = (str || "").trim();
  if (!s || s.length < 3)
    return "El nombre completo debe tener al menos 3 caracteres.";
  if (!REGEX_SOLO_LETRAS.test(s))
    return "Solo se permiten letras, espacios y guiones. No se permiten números ni caracteres especiales.";
  if (!tieneSentido(s))
    return "El nombre no parece válido. Evite escribir letras al azar o sin sentido.";
  return null;
};

// ─────────────────────────────────────────────────────────────────────────────
// TELÉFONO COLOMBIANO
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Valida un número celular colombiano.
 * Reglas: solo dígitos, empieza por 3, exactamente 10 dígitos.
 */
export const validarTelefono = (str) => {
  const s = (str || "").trim();
  if (!s) return "El teléfono es obligatorio.";
  if (!/^\d+$/.test(s))
    return "El teléfono solo puede contener números, sin espacios ni guiones.";
  if (!s.startsWith("3"))
    return "El número celular colombiano debe empezar por 3 (ej: 3001234567).";
  if (s.length !== 10)
    return "El número celular debe tener exactamente 10 dígitos.";
  return null;
};

// ─────────────────────────────────────────────────────────────────────────────
// CORREO ELECTRÓNICO
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Valida un correo electrónico con formato estándar.
 */
export const validarEmail = (str) => {
  const s = (str || "").trim();
  if (!s) return "El correo electrónico es obligatorio.";
  const regex = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
  if (!regex.test(s))
    return "El correo electrónico no es válido. Ejemplo: usuario@gmail.com";
  // Verifica que no sea solo @algo.x sin usuario real
  const local = s.split("@")[0];
  if (local.length < 2)
    return "La parte del usuario del correo es demasiado corta.";
  return null;
};

// ─────────────────────────────────────────────────────────────────────────────
// NÚMERO DE DOCUMENTO (según tipo)
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Tipos de documento y sus reglas:
 *  CC  (1) → solo dígitos, 6-10 caracteres
 *  CE  (2) → solo dígitos, 4-20 caracteres
 *  PA  (3) → Pasaporte: alfanumérico, máx 2 letras, 4-20 caracteres
 *  PEP (4) → solo dígitos, 4-20 caracteres
 *  PPT (5) → solo dígitos, 4-20 caracteres
 *
 * @param {string} str        Valor del documento
 * @param {number|string} tipoId  Valor del tipoDocumentoId
 * @param {string} tipoNombre Nombre del tipo (para mensajes). Opcional.
 */
export const validarDocumento = (str, tipoId, tipoNombre = "") => {
  const s = (str || "").trim();
  const id = Number.parseInt(tipoId, 10);
  const etiqueta = tipoNombre || `tipo ${tipoId}`;

  if (!s) return "El número de documento es obligatorio.";

  // CC: solo números, 6-10 dígitos
  if (id === 1) {
    if (!/^\d{6,10}$/.test(s))
      return "La Cédula de Ciudadanía (CC) debe tener entre 6 y 10 dígitos numéricos, sin letras ni caracteres especiales.";
    return null;
  }

  // Pasaporte (id 3): alfanumérico, máximo 2 letras
  if (id === 3) {
    if (!/^[a-zA-Z0-9]+$/.test(s))
      return `El documento (${etiqueta}) solo puede contener letras y números, sin caracteres especiales.`;
    const letras = (s.match(/[a-zA-Z]/g) || []).length;
    if (letras > 2)
      return `El pasaporte no puede tener más de 2 letras. Formato esperado: hasta 2 letras seguidas de números (ej: AB1234567).`;
    if (s.length < 4 || s.length > 20)
      return `El documento (${etiqueta}) debe tener entre 4 y 20 caracteres.`;
    return null;
  }

  // CE, PEP, PPT (ids 2, 4, 5): solo dígitos
  if ([2, 4, 5].includes(id)) {
    if (!/^\d+$/.test(s))
      return `El documento (${etiqueta}) debe contener solo dígitos, sin letras ni caracteres especiales.`;
    if (s.length < 4 || s.length > 20)
      return `El documento (${etiqueta}) debe tener entre 4 y 20 caracteres.`;
    return null;
  }

  // Caso genérico: solo dígitos
  if (!/^\d+$/.test(s))
    return "El número de documento debe contener solo dígitos, sin letras ni caracteres especiales.";
  return null;
};

// ─────────────────────────────────────────────────────────────────────────────
// TRANSPORTADORA
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Valida el nombre de una transportadora.
 * Acepta cualquier nombre de la lista TRANSPORTADORAS_CO o nombres
 * que solo contengan letras, números conocidos (472) y espacios.
 * No admite números aleatorios ni caracteres especiales.
 */
export const validarTransportadora = (str) => {
  const s = (str || "").trim();
  if (!s) return "La transportadora es obligatoria.";

  // Si es exactamente una de las transportadoras conocidas → OK
  const conocida = TRANSPORTADORAS_CO.map((t) => t.toLowerCase());
  if (conocida.includes(s.toLowerCase())) return null;

  // Regla general: solo letras, espacios, números y guiones/puntos
  // (ej: "472" es válido, "4x7z" no debería serlo)
  // Rechaza nombres puramente numéricos que no estén en la lista
  const soloNumeros = /^\d+$/.test(s);
  if (soloNumeros && !["472"].includes(s))
    return "La transportadora no puede ser solo un número. Escriba el nombre completo (ej: Servientrega, Coordinadora, 472).";

  if (!/^[a-zA-ZáéíóúüñÁÉÍÓÚÜÑ0-9\s'.,-]+$/.test(s))
    return "La transportadora no puede contener caracteres especiales como @, #, !, etc.";

  if (s.length < 2) return "El nombre de la transportadora es demasiado corto.";

  // Verificar que si tiene letras, estas tengan sentido
  const soloLetras = s.replaceAll(/[^a-zA-ZáéíóúüñÁÉÍÓÚÜÑ]/g, "");
  if (soloLetras.length > 2 && !tieneSentido(s))
    return "El nombre de la transportadora no parece válido. Evite escribir letras al azar.";

  return null;
};

File: 6_Trimestre/comunidad_inteligente/lib/utils/validaciones.dart
/// Utilidades de validación para formularios - Versión móvil
/// Portadas desde la versión web (validaciones.js)
library;

import 'package:flutter/services.dart';

// ══════════════════════════════════════════════════════════════════
// CONSTANTES
// ══════════════════════════════════════════════════════════════════

/// Patrón para nombres: letras (con tildes/ñ), espacios, apóstrofos, guiones
final RegExp _nombrePattern =
    RegExp(r"^[a-zA-ZáéíóúÁÉÍÓÚñÑüÜ\s'\-]+$");

/// Patrón para documentos numéricos (CC, CE, PEP, PPT)
final RegExp _soloDigitos = RegExp(r'^\d+$');

/// Patrón para pasaporte: alfanumérico, máximo 2 letras
final RegExp _pasaportePattern = RegExp(r'^[a-zA-Z0-9]+$');

/// Vocales para chequeo de sentido
final RegExp _vocales = RegExp(r'[aeiouáéíóúAEIOUÁÉÍÓÚ]');

/// Lista de transportadoras conocidas en Colombia
const List<String> transportadorasCo = [
  'servientrega',
  'interrapidisimo',
  'coordinadora',
  'deprisa',
  'tcc',
  'envia',
  'saferbo',
  'envía',
  'fedex',
  'dhl',
  'ups',
  '472',
  'tempo express',
  'domina',
  'redetrans',
  'aerocarga',
  'colvanes',
  'rapidísimo',
  'amazon',
  'mercado libre',
  'mercadolibre',
  'rappi',
  'uber',
  'falabella',
  'éxito',
  'exito',
  'linio',
  'shein',
  'temu',
  'aliexpress',
  'wish',
  'otro',
  'personal',
  'particular',
];

// ══════════════════════════════════════════════════════════════════
// INPUT FORMATTERS (filtros en tiempo real)
// ══════════════════════════════════════════════════════════════════

/// Filtro de entrada para nombres: solo letras, espacios, apóstrofos y guiones
class NombreInputFormatter extends TextInputFormatter {
  @override
  TextEditingValue formatEditUpdate(
    TextEditingValue oldValue,
    TextEditingValue newValue,
  ) {
    if (newValue.text.isEmpty) return newValue;
    // Permitir solo letras (con tildes/ñ), espacios, apóstrofos, guiones
    final filtered = newValue.text.replaceAll(
      RegExp(r"[^a-zA-ZáéíóúÁÉÍÓÚñÑüÜ\s'\-]"),
      '',
    );
    if (filtered == newValue.text) return newValue;
    return TextEditingValue(
      text: filtered,
      selection: TextSelection.collapsed(offset: filtered.length),
    );
  }
}

/// Filtro de entrada para documentos numéricos (CC, CE, PEP, PPT)
class DocumentoNumericoInputFormatter extends TextInputFormatter {
  @override
  TextEditingValue formatEditUpdate(
    TextEditingValue oldValue,
    TextEditingValue newValue,
  ) {
    if (newValue.text.isEmpty) return newValue;
    final filtered = newValue.text.replaceAll(RegExp(r'[^0-9]'), '');
    if (filtered == newValue.text) return newValue;
    return TextEditingValue(
      text: filtered,
      selection: TextSelection.collapsed(offset: filtered.length),
    );
  }
}

/// Filtro de entrada para pasaporte: alfanumérico, máximo 2 letras
class PasaporteInputFormatter extends TextInputFormatter {
  @override
  TextEditingValue formatEditUpdate(
    TextEditingValue oldValue,
    TextEditingValue newValue,
  ) {
    if (newValue.text.isEmpty) return newValue;
    // Solo alfanumérico
    String filtered = newValue.text.replaceAll(RegExp(r'[^a-zA-Z0-9]'), '');
    // Máximo 2 letras
    final letras = filtered.replaceAll(RegExp(r'[0-9]'), '');
    if (letras.length > 2) {
      filtered = oldValue.text;
    }
    if (filtered == newValue.text) return newValue;
    return TextEditingValue(
      text: filtered,
      selection: TextSelection.collapsed(offset: filtered.length),
    );
  }
}

/// Filtro de entrada para teléfono colombiano: solo dígitos, máx 10
class TelefonoInputFormatter extends TextInputFormatter {
  @override
  TextEditingValue formatEditUpdate(
    TextEditingValue oldValue,
    TextEditingValue newValue,
  ) {
    if (newValue.text.isEmpty) return newValue;
    String filtered = newValue.text.replaceAll(RegExp(r'[^0-9]'), '');
    if (filtered.length > 10) filtered = filtered.substring(0, 10);
    if (filtered == newValue.text) return newValue;
    return TextEditingValue(
      text: filtered,
      selection: TextSelection.collapsed(offset: filtered.length),
    );
  }
}

/// Retorna el InputFormatter adecuado según el tipo de documento
/// [tipoDocId] es el ID numérico del tipo de documento
/// Pasaporte = 2, los demás son numéricos
TextInputFormatter getDocumentoFormatter(int? tipoDocId) {
  if (tipoDocId == 2) return PasaporteInputFormatter();
  return DocumentoNumericoInputFormatter();
}

// ══════════════════════════════════════════════════════════════════
// VALIDATORS (para Form / TextFormField)
// ══════════════════════════════════════════════════════════════════

/// Verifica que un texto tenga "sentido" (al menos 15% vocales)
bool _tieneSentido(String str) {
  if (str.length < 3) return true; // Texto muy corto, permitir
  final totalVocales = _vocales.allMatches(str).length;
  return totalVocales / str.length >= 0.15;
}

/// Valida un nombre individual (primer nombre, primer apellido, etc.)
/// Retorna null si es válido, o el mensaje de error
String? validarNombre(String? value, {bool obligatorio = true}) {
  if (value == null || value.trim().isEmpty) {
    return obligatorio ? 'Campo requerido' : null;
  }
  final str = value.trim();
  if (str.length < 2) return 'Mínimo 2 caracteres';
  if (str.length > 45) return 'Máximo 45 caracteres';
  if (!_nombrePattern.hasMatch(str)) {
    return 'Solo se permiten letras, espacios y guiones';
  }
  if (!_tieneSentido(str)) {
    return 'El texto no parece un nombre válido';
  }
  return null;
}

/// Valida un nombre completo (campo de nombre concatenado)
String? validarNombreCompleto(String? value, {bool obligatorio = true}) {
  if (value == null || value.trim().isEmpty) {
    return obligatorio ? 'Campo requerido' : null;
  }
  final str = value.trim();
  if (str.length < 3) return 'Mínimo 3 caracteres';
  if (str.length > 100) return 'Máximo 100 caracteres';
  if (!_nombrePattern.hasMatch(str)) {
    return 'Solo se permiten letras, espacios y guiones';
  }
  if (!_tieneSentido(str)) {
    return 'El texto no parece un nombre válido';
  }
  return null;
}

/// Valida número de documento según tipo
/// tipoDocId: 1=CC, 2=PA, 3=CE, 4=PEP, 5=PPT (ajustar según BD)
String? validarDocumento(String? value, int? tipoDocId) {
  if (value == null || value.trim().isEmpty) return 'Campo requerido';
  final str = value.trim();

  switch (tipoDocId) {
    case 1: // Cédula de ciudadanía
      if (!_soloDigitos.hasMatch(str)) return 'Solo números para CC';
      if (str.length < 6 || str.length > 10) return 'CC debe tener entre 6 y 10 dígitos';
      break;
    case 2: // Pasaporte
      if (!_pasaportePattern.hasMatch(str)) return 'Solo letras y números';
      final letras = str.replaceAll(RegExp(r'[0-9]'), '');
      if (letras.length > 2) return 'Máximo 2 letras en pasaporte';
      if (str.length < 4 || str.length > 20) return 'Pasaporte: entre 4 y 20 caracteres';
      break;
    case 3: // Cédula de extranjería
    case 4: // PEP
    case 5: // PPT
      if (!_soloDigitos.hasMatch(str)) return 'Solo números para este documento';
      if (str.length < 4 || str.length > 20) return 'Entre 4 y 20 dígitos';
      break;
    default:
      if (str.length < 4 || str.length > 20) return 'Entre 4 y 20 caracteres';
  }
  return null;
}

/// Valida teléfono colombiano
String? validarTelefono(String? value, {bool obligatorio = false}) {
  if (value == null || value.trim().isEmpty) {
    return obligatorio ? 'Campo requerido' : null;
  }
  final str = value.trim();
  if (!_soloDigitos.hasMatch(str)) return 'Solo números';
  if (!str.startsWith('3')) return 'Debe iniciar con 3';
  if (str.length != 10) return 'Debe tener exactamente 10 dígitos';
  return null;
}

/// Valida correo electrónico
String? validarEmail(String? value, {bool obligatorio = false}) {
  if (value == null || value.trim().isEmpty) {
    return obligatorio ? 'Campo requerido' : null;
  }
  final str = value.trim();
  if (str.length > 100) return 'Máximo 100 caracteres';
  final emailRegex = RegExp(r'^[\w\-.]+@([\w\-]+\.)+[\w\-]{2,4}$');
  if (!emailRegex.hasMatch(str)) return 'Correo electrónico no válido';
  final parteLocal = str.split('@').first;
  if (parteLocal.length < 2) return 'La parte antes de @ debe tener al menos 2 caracteres';
  return null;
}

/// Valida transportadora (paquetería)
String? validarTransportadora(String? value) {
  if (value == null || value.trim().isEmpty) return 'Campo requerido';
  final str = value.trim();
  if (str.length < 2) return 'Mínimo 2 caracteres';
  if (str.length > 50) return 'Máximo 50 caracteres';
  if (RegExp(r'[^a-zA-ZáéíóúÁÉÍÓÚñÑüÜ0-9\s\-\.]').hasMatch(str)) {
    return 'Caracteres no permitidos';
  }
  if (!_tieneSentido(str)) {
    return 'El texto no parece una transportadora válida';
  }
  return null;
}

/// Valida contraseña
String? validarPassword(String? value) {
  if (value == null || value.trim().isEmpty) return 'Campo requerido';
  if (value.length < 6) return 'Mínimo 6 caracteres';
  if (value.length > 50) return 'Máximo 50 caracteres';
  return null;
}
